### DISCUSSION TITLE
Should we translate all the technical terms to Vietnamese?

### DISCUSSION THREAD

**Discussion by l0ngvh** [google/comprehensive-rust#2062] [General]
In my opinion, we shouldn't translate all technical terms from English to Vietnamese. Terms like `struct`, `implement`, `trait`, or type names such as `Deref` or `Vector` don't have Vietnamese counterparts that can accurately express them. Therefore, it's better to keep the English names to avoid confusion for the reader. If readers want to learn more about a technical term, they can do a quick search using the original English terms instead of having to translate them back from Vietnamese. This approach is often used for terms like `class` in other programming languages.

**Comment by l0ngvh** [comment]
cc @qu-ngx @daivinhtran

**Comment by l0ngvh** [comment]
For reference, [Rust Tiếng Việt](https://rust-tieng-viet.github.io) also doesn't translate many technical terms

**Reply by ghost** [reply]
Firstly, you are referring to a book. But you don't also look at how they translate some certain terms. Look at this one for example: 
<img width="812" alt="image" src="https://github.com/google/comprehensive-rust/assets/91497379/a53ada51-2ad7-4795-b31e-593b47f6f294">
This is how we are trying to do: To provide the meaningful Vietnamese translation for this course but not deviate from the original version. 

Secondly, if you take a look at the book introduction, it says: 
<img width="1280" alt="image" src="https://github.com/google/comprehensive-rust/assets/91497379/19cb3d8c-5b78-4bc4-964a-7955541a0ecc">

The point of this is to provide some notes with both English and Vietnamese. So is it purely for translation? I guess not. So are we providing notes with a partially translated English or are we doing translation?

**Reply by l0ngvh** [reply]
> To provide the meaningful Vietnamese translation for this course but not deviate from the original version.

I agree with you on this point. Our goal is to provide a meaningful Vietnamese translation while staying true to the original version. 

However, the purpose of this book is to help beginners get started with Rust, and it should enable readers to begin working with Rust after completing the course. When discussing how to implement something in Rust, readers should be able to refer to the correct Rust terms clearly. To achieve this, we might need to make some compromises in the translation process. These compromises are necessary because attempting to translate certain terms could potentially deviate from the original meaning.

In conclusion, I don't think doing a pure translation is a good idea.

**Reply by ghost** [reply]
@vohoanglong0107  Read my reply below. If you are not satisfied with my comments, I will just leave it to @daivinhtran. He might take even more time reviewing your translation. I think his comments will be the same with some untranslated parts if not all of them as he did take a lot of consideration with other PRs as well. You are free to look at the other merged earlier PRs.

**Reply by l0ngvh** [reply]
Understood. I just want us to reach a consensus on the best approach to translating this book moving forward, so that it can serve as a guideline for future potential contributors.

**Comment by ghost** [comment]
@vohoanglong0107 I agree with you that some certain built-in terms are too hard to translate. You can skip some of them. But make sure you read your translation so you don't feel it is too odd for other readers. You might be familiar with that way but others might not. I commented on the ones I feel odd tho. It feels like you should have something there instead of leaving it not translated like that. For example, "we implement something?" in that sentence is not the name of any built-in, then it should have some translations with the word "implement"?

**Reply by l0ngvh** [reply]
Wouldn't `implement` also be considered built-in? This term was adopted by Rust to mean something akin to `inheritance`, similar to terms like `struct` and `closures`. I don't even know how to translate `closures` 😅

**Reply by ghost** [reply]
That implement in the sentence is not the implement you are referring to btw.

**Reply by ghost** [reply]
`implement` (Rust built-in) != implement (an english word)

**Reply by l0ngvh** [reply]
I understand, then when `implement` is used as an English word, I will translate it. However, do you agree that we should keep terms like `String implement Deref` unchanged?

**Reply by ghost** [reply]
@vohoanglong0107 I agree. But let me see if I put some wrong comments on your translations first. Also try to add translation to other terms as well if possible. If you are unsure or feel it wierd then just put the Vietnamese version in brackets.

**Reply by l0ngvh** [reply]
> Also try to add translation to other terms as well if possible

Okay!

> If you are unsure or feel it wierd then just put the Vietnamese version in brackets.

@qu-ngx Regarding this, I'm not sure if we should overuse brackets like that. I propose we add brackets the first time we encounter the terms, such as at the start of a chapter or session.

**Reply by ghost** [reply]
@vohoanglong0107 Sure. It also works!

**Comment by l0ngvh** [comment]
I think this is also a good place to discuss the translation of some common terms that could potentially be translated into Vietnamese:

| Original | Translated |
|---|---|
| generic | tổng quát (?) |

We will populate this table the more we translate the book.

**Reply by ghost** [reply]
I guess not. I am afraid there will be something conflicting the docs generation. Just put that one in brackets I think? For example:

loại tổng quát / loại chung (generic type)

This should be good. I am just only afraid that you might not pass some test cases if you add more contents than they allow you to.

**Reply by l0ngvh** [reply]
@qu-ngx What are your thoughts on translating `generic`? A `generic type` could be translated into Vietnamese as `kiểu dữ liệu tổng quát`. I prefer to use the term `generic` as is, but I would like to hear your opinion on this.

**Reply by ghost** [reply]
yea "kiểu dữ liệu tổng quát" is good translation for it.

**Reply by l0ngvh** [reply]
> This should be good. I am just only afraid that you might not pass some test cases if you add more contents than they allow you to.

I'm unaware of any content-based limit tests in the repository. There's only a line limit, but we can simply run dprint fmt, and it will automatically format the content accordingly.

**Reply by tidefield** [reply]
Hi @qu-ngx @vohoanglong0107 . Thanks for the discussion here and in the other PRs. I appreciate the arguments. Let's hold on merging the PRs. I have a few thoughts and will write up a throughout comment so we have some ground for discussion.

**Reply by mgeisler** [reply]
The idea of a translation table is partly why we have this page: [Glossary](https://google.github.io/comprehensive-rust/glossary.html).

The page was started some time ago and haven't been updated much since — if you find words missing, please add them!

> > This should be good. I am just only afraid that you might not pass some test cases if you add more contents than they allow you to.
> 
> I'm unaware of any content-based limit tests in the repository. There's only a line limit, but we can simply run `dprint fmt`, and it will automatically format the content accordingly

Correct, we don't have any automated tests for the length of text in the English or other versions.

Since we regularly use the English version for classroom training at Google and elsewhere, people will notice if the slides become too big — if people start using the translated versions regularly for classroom training, then they should contact the translators and talk about how the text can be reformulated in a short and precise way. This is super hard, so I would not worry too much about it right now.

**Comment by tidefield** [comment]
I can see this can be a (controversial) debate and this debate is not new in other languages. See https://github.com/google/comprehensive-rust/discussions/1184.

I think you both (@vohoanglong0107 @qu-ngx ) agree it doesn't read naturally if `struct`, `closures`, `test`, `string` are translated to Vietnamese. In fact, it's a little counter-intuitive because the English words are more familiar to developers in this context. So we can expand this list (starting with `struct`, `closures`, `test`, `string`) to all the terms that we want to keep as is.

Per https://github.com/google/comprehensive-rust/discussions/1184#discussioncomment-7250542, they (cc @mgeisler) introduced [this glossary](https://google.github.io/comprehensive-rust/glossary.html) as an escape hatch (sort of) to this debate.

Here is how I think we should move forward:

1. Start translating https://google.github.io/comprehensive-rust/glossary.html now. This includes `generic` in the table @vohoanglong0107 mentioned above.
2. As you translate and there are more terms that you want to keep as is, add them to https://google.github.io/comprehensive-rust/glossary.html (the English version) and then translate them to Vietnamese. When a Vietnamese reader encounters an English term, they should be able to find a translation in the translated glossary.
3. Code block: I think we should just keep all code block as is for now except for comments because comments are usually meant to explain the meaning/context behind the code.

This is some consensus we can reach. I'm open to modifications. 

One thing to keep in mind is that we don't need to get everything perfectly right now. We can translate the book in a way that has room for improvement later. If we translate all of the terms, a term can lead to more than one translation which can cause divergence that readers find hard to grasp; it's also hard to bring all the divergences back to a central Vietnamese term.

**Reply by l0ngvh** [reply]
I'm on board with this idea. We could discuss more details about each translation term during the translation of the glossary.

**Reply by mgeisler** [reply]
Hi all, I added a comment above saying basically this :smile: So I fully support the idea of using the glossary as a tool for keeping some minimum amount of consistency.

Translating is super hard and it won't be perfect on the first few tries... the course itself is also not perfect and keeps evolving here and there. So iteration it key in my opinion.

**Comment by tidefield** [comment]
By the way, https://google.github.io/comprehensive-rust/vi/index.html is published so we can see all the translated content.
